### ISSUE TITLE
Database purged periodically?

### ISSUE THREAD

**Issue by fellnerse** [ajeetdsouza/zoxide#433]
Hey, I love your tool so far, it works really great with nushell.

But somehow the database is purged periodically. So basically over night the `db.zo` file in the `zoxide` is missing and it does not remember anything from the day before.

Did you experience similar behaviour? Do you have any idea of how to solve this? When I print out the contents of the file after changing to this folder I can see that it correctly added the path to the folder there, but everything else is gone.

Thanks in advance!

**Comment by ajeetdsouza** [comment]
I haven't experienced anything like that, unfortunately.

- What version of zoxide are you on?
- What OS are you on?
- How often does this issue occur?
- Is there any other file inside the data directory?

**Comment by fellnerse** [comment]
version: v0.8.2
macOs: 12.4 M1
I noticed it now always the first time im using my laptop again in the morning (and it also actually happened right now after lunch break, feels kind of random)
No the directory is emtpy, I did not set the environment variable for chaning the directoyr

**Comment by ajeetdsouza** [comment]
I use zoxide on macOS 12 too, never had this problem before. Database writes on UNIX systems are atomic, so I'm not sure why this is happening. Could you try changing `$_ZO_DATA_DIR` to `$HOME` and seeing if it still occurs?

**Comment by fellnerse** [comment]
Then it probably is related to something else not directly `zoxide`. Did you ever try the `nushell` integration?
What I do not understand is why I have to add `zoxide init nushell --hook prompt --cmd j | save ~/.zoxide.nu` to the `env.nu` file. This should always create the same file over and over again, right? I thought it might be enough to just source ` ~/.zoxide.nu`

Will try with a different folder!

**Comment by ajeetdsouza** [comment]
> What I do not understand is why I have to add `zoxide init nushell --hook prompt --cmd j | save ~/.zoxide.nu` to the `env.nu` file.

Nushell hasn't had a stable release yet, so each release keeps breaking zoxide. Adding that line to your `env.nu` ensures that your initialization script always stays updated when you update zoxide. However, it will work just as well if you manually run this command every time you update zoxide.

**Comment by ajeetdsouza** [comment]
Hey @fellnerse, are you still experiencing this issue?

**Comment by fellnerse** [comment]
@ajeetdsouza Yes sadly, I think it is reset after restart for sure, but also at other times as well. Putting the database in my homefolder did not help :/
Is there any way I can debug this?

**Comment by ajeetdsouza** [comment]
zoxide doesn't run anything on system startup, so it's very strange that your database file is always deleted upon restart. What I would do to debug this is move the zoxide binary to another location and replace it with a simple wrapper Python script:

```python
import os.path
import subprocess
import sys

db_path = "<location of zoxide DB>"
bin_path = "<location of actual zoxide binary>"
log_path = "/home/foo/zoxide.log"

with open(log_path, "a") as f:
  # error out if DB exists
  if not os.path.exists(db_path):
    f.write("zoxide could not start, DB is missing\n")
    sys.exit(1)

  # run the actual zoxide binary
  argv = sys.argv
  argv[0] = bin_path
  subprocess.run(argv)
  
  # error out if DB exists
  if not os.path.exists(db_path):
    f.write("zoxide deleted the database\n")
    sys.exit(1)
```

After this I'd try to restart and see if there's any output in the logfile.

**Comment by fellnerse** [comment]
Thanks for the script, I integrated it and got the following results:
![image](https://user-images.githubusercontent.com/41167496/187164214-72b1fbcb-d38e-4144-819b-4a57e2b24d6c.png)
Between the two cats on the db file I waited over night; as you can see the log is still empty (I checked that it worked, it logs when the file is misssing completly). But somehow it seems that its purged.
<img width="293" alt="image" src="https://user-images.githubusercontent.com/41167496/187165665-1568252f-584a-4714-a6f5-e50b60496a01.png">

**Comment by ajeetdsouza** [comment]
Hmm, seems like your database has been purged but not deleted. Can you try replacing both instances of `if not os.path.exists(db_path)` with `if os.path.getsize(db_path) < 16`?

**Comment by ajeetdsouza** [comment]
It's odd that you're not seeing any updates. Are you checking the correct _ZO_DATA_DIR? Does `zoxide add .` work?

**Comment by fellnerse** [comment]
Yes the path is correct, as it is working sometimes :D and `zoxide add .` seems not to work:
![image](https://user-images.githubusercontent.com/41167496/187610362-77a84dbc-a1ed-4e02-b1cb-8f94186d467a.png)

**Comment by ajeetdsouza** [comment]
`~` does not get added to the database, because it can be accessed by `z ~` or simply `z`. It should work for other directories.

**Comment by fellnerse** [comment]
You are right, that does work; but still the problem is that the db is purged periodically, and I do not know why. I think we can rule out deletion at that point.

**Comment by ajeetdsouza** [comment]
Did you try the script with `os.path.getsize()`?

**Comment by fellnerse** [comment]
Yes, 16 seemed too small of a value so I used 150 (not sure what the unit is tho). But do not get any logs

**Comment by ajeetdsouza** [comment]
As odd as it seems, if you're not getting logs, it seems like zoxide is not the thing deleting your data?

**Comment by fellnerse** [comment]
Well the db-file also is not deleted. I highly doubt that I have process that interacts with the database and executes a drop table.
But well thanks for your help, I guess I will use a different tool

**Comment by ajeetdsouza** [comment]
True, but I have no data to work with here:

- The issue is unreproducible on my M1 MacBook running Nushell
- There's never been a report of data being purged in this manner
- We've added checks for every invocation of zoxide, and there are no logs indicating that zoxide is deleting data
- The issue occurs when you restart your machine, but zoxide is a shell plugin and doesn't do anything on system restart

As a last resort, I'd suggest two things:

- Check that `_ZO_MAXAGE` isn't being set anywhere. If this is set to `0`, it could cause the deleting issue you described
- I've just released v0.8.3, where I've changed the initialization method for Nushell systems, maybe that fixes your issue
